### PR TITLE
MT Head

### PR DESCRIPTION
**Pull request by PhakphumAdev**
- Implementation of MT Head with MBart as helper model 
- run name for each experiment name in tensorboard

### CODE REVIEW

**Comment by lgessler** [comment]
LGTM! Just one thing, can you rename `ud_mt.py` to `mt.py`? (The UD there stands for Universal Dependencies, which I included because the POS and parser heads both rely on UD data. But the MT head doesn't.)

File: microbert2/microbert/tasks/ud_mt.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+            "test":  read_parallel_tsv(test_mt_path, delimiter) if test_mt_path else [],
+        }
+        self._proportion = proportion
+        self._mbart_tokenizer_name = mbart_tokenizer_name
+        self._tgt_lang_code = tgt_lang_code
+
+        # MBART 
+        self._tok = AutoTokenizer.from_pretrained(mbart_tokenizer_name,use_fast=False)
+        #self._tok.add_special_tokens({"additional_special_tokens": ["<cop_XX>"]})
+        self._tok.src_lang = "ar_AR"
```
Comment (lgessler) [inline microbert2/microbert/tasks/ud_mt.py]: This should be a parameter instead of a hardcoded "ar_AR"

File: microbert2/microbert/tasks/ud_mt.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+class MTTask(MicroBERTTask, CustomDetHash):
+
+    def __init__(
+            self,
+            head: Lazy[MTHead],
+            train_mt_path: str,
+            dev_mt_path: str,
+            test_mt_path: Optional[str] = None,
+            delimiter: str = "\t",
+            proportion: float = 0.1, #0.2 0.5 coptic
```
Comment (lgessler) [inline microbert2/microbert/tasks/ud_mt.py]: Remove the comment here

File: microbert2/microbert/tasks/ud_mt.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+            mbart_tokenizer_name: str = "facebook/mbart-large-50",
+            tgt_lang_code: str = "en_XX",
+            max_tgt_len: int = 128,
```
Comment (lgessler) [inline microbert2/microbert/tasks/ud_mt.py]: I think we might want to default to 512.

File: microbert2/microbert/tasks/ud_mt.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+logger = getLogger(__name__)
+
+class MTHead(torch.nn.Module, FromParams):
+    def __init__(
+            self,
+            num_layers: int,
+            embedding_dim: int,
+            mbert_model_name: str = "facebook/mbart-large-50",
+            use_layer_mix: bool = True,
```
Comment (lgessler) [inline microbert2/microbert/tasks/ud_mt.py]: Let's have this default to `False`.

File: microbert2/microbert/tasks/ud_mt.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+from microbert2.common import pool_embeddings
+from microbert2.microbert.tasks.task import MicroBERTTask
+
+logger = getLogger(__name__)
+
+class MTHead(torch.nn.Module, FromParams):
+    def __init__(
+            self,
+            num_layers: int,
```
Comment (lgessler) [inline microbert2/microbert/tasks/ud_mt.py]: This is a bit confusing right now because you might think it has to do with the head. I think the name `num_encoder_layers` would be better.

File: microbert2/microbert/tasks/ud_mt.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+            embedding_dim: int,
+            mbert_model_name: str = "facebook/mbart-large-50",
+            use_layer_mix: bool = True,
+            freeze_decoder: bool = True,
+            train_last_k_decoder_layers: int = 0
+    ):
+        super().__init__()
+        self.use_layer_mix = use_layer_mix
+        self.proj = None
+        self.perplexity = Perplexity(ignore_index=-100)
```
Comment (lgessler) [inline microbert2/microbert/tasks/ud_mt.py]: This is a potential correctness issue. I think you want to instead get the pad index from mbart like you do in lines 91-93 below. I don't know what `ignore_index` is correct for mbart, but I'd expect mbart to use `0` instead. Anyway, you can just take it directly from the model and not worry about hard-coding a value.

File: microbert2/microbert/tasks/ud_mt.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+        if self.proj is not None:
+            enc = self.proj(enc)
+        
+        enc_out = BaseModelOutput(last_hidden_state=enc)
+        #Mask pad tokens in labels so loss ignores them
+        labels = tgt_input_ids.clone()
+        pad_id = self.mbart.config.pad_token_id
+        if pad_id is not None:
+            labels[labels == pad_id] = -100
```
Comment (lgessler) [inline microbert2/microbert/tasks/ud_mt.py]: Same as above--we should avoid hardcoding any values if we can just get the value from the model config

File: microbert2/microbert/tasks/ud_mt.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        out = self.mbart(
+            encoder_outputs=enc_out,
+            attention_mask=encoder_attention_mask, 
+            decoder_attention_mask=tgt_attention_mask,
+            labels=labels,                 
+            use_cache=False,
+        )
+        loss = out.loss
+        log_probs = F.log_softmax(out.logits, dim=-1)
+        self.perplexity.update(log_probs, labels)
```
Comment (lgessler) [inline microbert2/microbert/tasks/ud_mt.py]: I'm pretty sure you just want to feed `out.logits` here and not `log_probs`, right?

Comment (lgessler) [reply microbert2/microbert/tasks/ud_mt.py]: [Documentation](https://lightning.ai/docs/torchmetrics/stable/text/perplexity.html) says:


    * preds ([Tensor](https://docs.pytorch.org/docs/stable/tensors.html#torch.Tensor)): Logits or a unnormalized score assigned to each token in a sequence with shape [batch_size, seq_len, vocab_size], which is the output of a language model. Scores will be normalized internally using softmax.

**Review by lgessler — CHANGES_REQUESTED** [review]
Found some small things actually

**Comment by lgessler** [comment]
Looks good! I'm going to make a few more tweaks in the main branch.
